Alloy/Shaders/Shaders.metal

```diff
@@ -11,18 +11,186 @@
+
+    const ushort2 block_size_coef = ushort2(block_size.width, block_size.height);
+    float2 read_position = float2(thread_position_in_grid * block_size_coef);
+
+    constexpr sampler s(coord::pixel,
+                        address::clamp_to_zero,
+                        filter::nearest);
+
+    half4 min_value_in_block = input_texture.sample(s, read_position);
+
+    for (ushort x = 0; x < block_size.width; x++) {
+        for (ushort y = 1; y < block_size.height; y++) {
+
+            // Prevent going out of texture.
+            if (read_position.x >= input_texture_width || read_position.y >= input_texture_height) {
+                break;
+            }
+
+            read_position = float2(thread_position_in_grid * block_size_coef + ushort2(x, y));
+            const half4 current_value = input_texture.sample(s, read_position);
+            min_value_in_block = min(min_value_in_block, current_value);
+        }
+    }
+
+    shared_memory[thread_index_in_threadgroup] = min_value_in_block;
+
+    threadgroup_barrier(mem_flags::mem_threadgroup);
+
+    if (thread_index_in_threadgroup == 0) {
+
+        half4 min_value = shared_memory[0];
+        const ushort threads_in_threadgroup = threads_per_threadgroup.x * threads_per_threadgroup.y;
+        for (ushort i = 1; i < threads_in_threadgroup; i++) {
+            half4 min_value_in_block = shared_memory[i];
+            min_value = min(min_value, min_value_in_block);
+        }
+
+        result = float4(min_value);
+    }
+
+}
+
+kernel void mean(texture2d<half, access::sample> input_texture [[ texture(0) ]],
+                 constant BlockSize& block_size [[ buffer(0) ]],
+                 device float4& result [[ buffer(1) ]],
+                 threadgroup half4* shared_memory [[ threadgroup(0) ]],
+                 const ushort thread_index_in_threadgroup [[ thread_index_in_threadgroup ]],
+                 const ushort2 thread_position_in_grid [[ thread_position_in_grid ]],
+                 const ushort2 threads_per_threadgroup [[ threads_per_threadgroup ]]) {
+
+    const half input_texture_width = input_texture.get_width();
+    const half input_texture_height = input_texture.get_height();
+
+    const ushort2 block_size_coef = ushort2(block_size.width, block_size.height);
+    float2 read_position = float2(thread_position_in_grid * block_size_coef);
+
+    constexpr sampler s(coord::pixel,
+                        address::clamp_to_zero,
+                        filter::nearest);
+
+    half4 total_sum_in_block = input_texture.sample(s, read_position);
+
+    for (ushort x = 0; x < block_size.width; x++) {
+        for (ushort y = 1; y < block_size.height; y++) {
+
+            // Prevent going out of texture.
+            if (read_position.x >= input_texture_width || read_position.y >= input_texture_height) {
+                break;
+            }
+
+            read_position = float2(thread_position_in_grid * block_size_coef + ushort2(x, y));
```

Comment:
why recalculate each time if you already have read_position, btw better name is blockStart or something like that